File: lintcode/Median.cpp
class Solution {
public:
    /**
     * @param nums: A list of integers.
     * @return: An integer denotes the middle number of the array.
     */
    int median(vector<int> &nums) {
        // write your code here
        if(nums.empty()) throw runtime_error("wrong input!");
        
        int n = nums.size();
        if(n == 1) return nums[0];
        
        int k = (n+1)/2;
        return kthElem(nums, k, 0, n-1);
    }
private:
    int kthElem(vector<int> &A, int k, int s, int e){
        if(s > e) throw runtime_error("wrong range!");
        
        int p = partition(A, s, e);
        int q = p-s+1;
        
        if(q == k) return A[p];
        else if(q < k) return kthElem(A, k-q, p+1, e);
        else return kthElem(A, k, s, p-1);
    }
    
    int partition(vector<int> &A, int s, int e){
        int i = s;
        int j = e;
        
        int x = A[s];
        while(i <= j){
            if(A[i] <= x) i++;
            else{
                swap(A[i], A[j--]);
            }
        }
        swap(A[s], A[j]);
        
        return j;
    }
};

File: lintcode/java/Median.java
public class Solution {
    /**
     * @param nums: A list of integers.
     * @return: An integer denotes the middle number of the array.
     */
    public int median(int[] nums) {
        if(nums.length == 0) throw new IllegalArgumentException("xxx");
        if(nums.length == 1) return nums[0];
        
        int k = (nums.length+1)/2;
        return findKthElem(nums, 0, nums.length-1, k);
    }
    
    private int findKthElem(int[] A, int s, int e, int k){
        if(s > e) throw new IllegalArgumentException("xxx");
        
        int p = partation(A, s, e);
        int m = p-s+1;
        
        if(m == k) return A[p];
        else if(m < k) return findKthElem(A, p+1, e, k-m);
        else return findKthElem(A, s, p-1, k);
    }
    
    // [s, e]
    private int partation(int[] A, int s, int e){
        if(s > e) throw new IllegalArgumentException("xxxx");
        
        int p = A[s];
        int i = s+1;
        int j = e;
        
        while(i <= j){
            if(A[i] <= p) i++;
            else if(A[i] > p) swap(A, i, j--);
        }
        swap(A, s, j);
        return j;
    }
    
    private void swap(int[] A, int i, int j) {
        int t = A[i];
        A[i] = A[j];
        A[j] = t;
    }
}
